### PR TITLE
Add tw caldav sync

### PR DESCRIPTION
**Pull request by Pete-Hamlin**
# Description

First off wanted to just say this project rocks!

The goal of this PR is to expand on the google calendar sync to support arbitrary caldav servers (e.g. nextcloud) as a new supported service.
I don't personally use google, but really wanted a viable way to sync tasks on my phone to taskwarrior (I've given up on the hope of taskserver being fit for purpose).

I've gone for 'least disruptive' in terms of adding new features, so there is a lot of shared/similar code between this service and gcal sync.
It would probably be possible to refactor some of that down into shared modules, but wanted to kneep things decoupled at first glance.

This is my first PR, so please do feel free to tear it to shreds if it's not up to standard!
Equally if this feature isn't something you'd like to have included as part of the project, I'm happy to keep it entirely on my own fork.
Just figured it was something that was useful for me, and wanted to share that back upstream.

### New Dependencies

```
caldav = { version = "^0.11.0", optional = true }
icalendar = { version = "^5.0.3", optional = true }
```

Can probably get away with not directly requiring `icalendar` as it is a dependency of `caldav`.
However, as we're directly using it, I figured I should add it for future proofing.

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update (update included in PR)

# How Has This Been Tested?

Added a set of tests for the conversion method, inspired by the `gcal_tw` test suite.
They should run as part of the `pytest` suite.
Have also tested against (and am currently using with) my personal nextcloud instance.
For obvious reasons, I won't include the details of this in the PR, however I am happy to setup a test user/calendar on my server if you wish to test in a live environment.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Any dependent changes have been merged and published in downstream modules

## TODO

- [x] Rebase on top of latest master
- [x] Address series of comments from first PR pass
- [x] Verify tests pass
- [x] @bergercookie to give it a go locally and potentially leave more comments

### CODE REVIEW

**Comment by bergercookie** [comment]
Hi @Pete-Hamlin ,

Thanks for the kind words and this work. I've long wanted to have a caldav integration and this looks awesome! 🎉🎉🎉

I'm in the process of generalizing a bit `taskwarrior-syncall` (see work in this branch - https://github.com/bergercookie/syncall/tree/gkeep-fs). Once I'm through with that I'm super keen to review this and get it merged ASAP.

**Comment by bergercookie** [comment]
> Handle recurring tasks (will need to manually inject the appropriate ics format when adding/updating caldav tasks)

I find this a challenging task, given how taskwarrior manages recurring tasks. I would prefer if there was no assumption about the recurrence and rather just sync the next coming events during the next synchronization run.

File: taskwarrior_syncall/tw_caldav_utils.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+        caldav_item["start"] = tw_item["due"] - timedelta(hours=1)
+        caldav_item["due"] = tw_item["due"]
+
+    if "tags" in tw_item.keys():
+        caldav_item["categories"] = tw_item["tags"]
+
+    return caldav_item
+
+
+def convert_caldav_to_tw(caldav_item: Item) -> Item:
```
Comment (bergercookie) [inline taskwarrior_syncall/tw_caldav_utils.py:64]: Absolutely not critical and if too much of a pain ignore this comment  but, in my latest merge, I introduced a `ConcreteItem` class - and I'm using it instead of `Item` which basically is a dict. Maybe you could use that instead of instantiating and passing around an `Item` ?

https://github.com/bergercookie/syncall/blob/master/syncall/concrete_item.py

And I'm using it as the building block (instead of `Item`) for `FilesystemFile` and `FilesystemSide`

https://github.com/bergercookie/syncall/blob/master/syncall/filesystem/filesystem_file.py

https://github.com/bergercookie/syncall/blob/master/syncall/filesystem/filesystem_side.py

File: taskwarrior_syncall/tw_caldav_utils.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    if "due" in caldav_item.keys():
+        tw_item["due"] = caldav_item["due"]
+
+    if "categories" in caldav_item.keys():
+        tw_item["tags"] = caldav_item["categories"]
+
+    return tw_item
+
+
+def map_ics_to_item(vtodo) -> Dict:
```
Comment (bergercookie) [inline taskwarrior_syncall/tw_caldav_utils.py:111]: Maybe prefer `Mapping` over `Dict` ? https://stackoverflow.com/questions/52487663/python-type-hints-typing-mapping-vs-typing-dict

**Review by bergercookie — COMMENTED** [review]
Went through a first pass and left some comments.
If you've used the gcal_side as a guideline for this new integration, feel free to give it a second look as I've made quite a few changes to it in my recent merge.

> Consider how to refactor out extra steps in conversion, and just store caldav items in their vTodo formats (though this will make test files much uglier)

That's one thing I wanted to ask about actually. Is connecting to a caldav server using username/password the best strategy for this?

Could perhaps `tw_caldav_sync` synchronize the contents of the TW db with the contents of one/multiple icalendar files instead?

**Comment by bergercookie** [comment]
Overall looks good. Here's some extra TODOs before merging

- [ ] Rebase on top of latest master
- [ ] Addres series of comments from first PR pass
- [ ] Verify tests pass
- [ ] @bergercookie to give it a go locally and potentially leave more comments

**Comment by Pete-Hamlin** [comment]
Thanks for the review.

I'll try and make some time to look at addressing the issues over the weekend.


> I find this a challenging task, given how taskwarrior manages recurring tasks. I would prefer if there was no assumption about the recurrence and rather just sync the next coming events during the next synchronization run.

Tbh, I'm not overly familiar with how TW does recurring tasks, as that's not a feature I really use.
Happy to just scrub that idea in that case.

> That's one thing I wanted to ask about actually. Is connecting to a caldav server using username/password the best strategy for this?
> 
> Could perhaps `tw_caldav_sync` synchronize the contents of the TW db with the contents of one/multiple icalendar files instead?

As in how [vdirsyncer](https://github.com/pimutils/vdirsyncer) does it, or something different?

My main motivation was that I just wanted a single touch set & forget solution that would work with my nextcloud instance, and directly syncing with a caldav instance seemed the best way to do that.
Syncing with an icalendar file would require another 3rd party program to get that file locally from my server, or would require additional steps somewhere along the process.

I'm not sure if /what the established practice is when it comes to caldav, so I figured I'd just go with the method that best suits my usecase :)

**Comment by bergercookie** [comment]
> I'll try and make some time to look at addressing the issues over the weekend.

Cool, let me know if you're busy and I can also jump in to make some of these changes.
 
> As in how [vdirsyncer](https://github.com/pimutils/vdirsyncer) does it, or something different?

Yes, that's what I originally had in mind as well. Tbh, though, if this one works for you let's just stick with it. :)

File: taskwarrior_syncall/scripts/tw_caldav_sync.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+    format_dict,
+    log_to_syslog,
+    logger,
+    loguru_tqdm_sink,
+)
+
+try:
+    from taskwarrior_syncall import CaldavSide
+except ImportError:
+    inform_about_app_extras(["caldav-client"])
```
[the remark this answers is not in the document]
Comment (Pete-Hamlin) [reply taskwarrior_syncall/scripts/tw_caldav_sync.py]: Not sure I follow I'm afraid, have corrected to the following:
```python
try:
    from syncall import CaldavSide, TaskWarriorSide
except ImportError:
    inform_about_app_extras(["caldav", "tw"])
```
This is mirroring how it's been don in `asana` I think?

File: readme-caldav.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+With environment variables:
+
+```
+CALDAV_USER=myUser CALDAV_PASSWD=myPass tw_caldav_sync --caldav-url https://nextcloud.example.com/remote.php/dav --caldav-calendar MyCalendar --taskwarrior-tags sync
+```
+
+## Future Work
+
+- [ ] Handle recurring tasks (will need to manually inject the appropriate ics format when adding/updating caldav tasks)
```
[the remark this answers is not in the document]
Comment (Pete-Hamlin) [reply readme-caldav.md:84]: bc69eb50a90fcbaa18b5c85f6e1b5fdce7951acd

File: readme-caldav.md [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+```sh
+pip3 install taskwarrior-syncall[caldav-client]
+```
+
+## Usage
+
+### Taskwarrior
+
+You can find synchronize a series of Taskwarrior tasks that have a particular
```
[the remark this answers is not in the document]
Comment (Pete-Hamlin) [reply readme-caldav.md:47]: bc69eb50a90fcbaa18b5c85f6e1b5fdce7951acd

File: readme-caldav.md [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+- Tw `tags` <-> `CATEGORIES`
+
+### Current limitations
+
+- No specific support for "waiting" tasks in taskwarrior, they will be treated like any other "needs-action" caldav task
+- No support for recurring tasks sync in either direction
+
+## Installation
+
+Install the `taskwarrior-syncall` package from PyPI, enabling the `caldav-client`
```
[the remark this answers is not in the document]
Comment (Pete-Hamlin) [reply readme-caldav.md:36]: bc69eb50a90fcbaa18b5c85f6e1b5fdce7951acd

File: taskwarrior_syncall/scripts/tw_caldav_sync.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Pete-Hamlin) [reply taskwarrior_syncall/scripts/tw_caldav_sync.py]: bc69eb50a90fcbaa18b5c85f6e1b5fdce7951acd

File: taskwarrior_syncall/scripts/tw_caldav_sync.py [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+                "tw_project": tw_project,
+                "tw_tags": tw_tags,
+            },
+            config_fname="tw_caldav_configs",
+            custom_combination_savename=custom_combination_savename,
+        )
+
+    # at least one of tw_tags, tw_project should be set ---------------------------------------
+    if not tw_tags and not tw_project:
+        raise RuntimeError(
```
[the remark this answers is not in the document]
Comment (Pete-Hamlin) [reply taskwarrior_syncall/scripts/tw_caldav_sync.py]: 4bc702c8bea43e4004bc87d12f0fed4c356bef7f

File: taskwarrior_syncall/scripts/tw_caldav_sync.py [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+            "You must provide a URL and calendar in order to synchronize via caldav"
+        )
+
+    # fetch username
+    if not caldav_user:
+        caldav_user = os.environ.get("CALDAV_USERNAME")
+    if caldav_user is None:
+        raise RuntimeError("You must provide a username in order to synchronize via caldav")
+    assert caldav_user
```
[the remark this answers is not in the document]
Comment (Pete-Hamlin) [reply taskwarrior_syncall/scripts/tw_caldav_sync.py]: 4bc702c8bea43e4004bc87d12f0fed4c356bef7f

File: taskwarrior_syncall/caldav_side.py [hunk 9]
```diff
@@ -0,0 +1,142 @@
+from typing import Dict, Sequence
```
[the remark this answers is not in the document]
Comment (Pete-Hamlin) [reply taskwarrior_syncall/caldav_side.py]: bc69eb50a90fcbaa18b5c85f6e1b5fdce7951acd

File: taskwarrior_syncall/scripts/tw_caldav_sync.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Pete-Hamlin) [reply taskwarrior_syncall/scripts/tw_caldav_sync.py]: OK nevermind,  I've got it now, fixed in 012ddc00a96072703cd81a547105876318dca087

**Comment by Pete-Hamlin** [comment]
OK, have merged with `main` and have started going through some of the issues.

Will take another stab when I get the time.
If you'd like address any of the remaining issues, by all means feel free :)

If not, I'll try and get them sorted by the end of the week

**Comment by bergercookie** [comment]
Hi @Pete-Hamlin . Looking at  the PR now, I may make some changes before end of day today :)

File: syncall/tw_caldav_utils.py [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+    # Iterate through rest of lines, find only the uuid
+    for line in lines[i:]:
+        parts = line.split(":", maxsplit=1)
+        if len(parts) == 2 and parts[0].lower().startswith("* uuid"):
+            try:
+                uuid = UUID(parts[1].strip())
+            except ValueError as err:
+                logger.error(
+                    f'Invalid UUID "{err}" provided during caldav -> TW conversion,'
+                    f" Using None...\n\n{traceback.format_exc()}"
```
Comment (bergercookie) [inline syncall/tw_caldav_utils.py:174]: What does it mean for the UUID to be None?

**Comment by Pete-Hamlin** [comment]
Hi @bergercookie,

Apologies for the relative radio silence, this last couple of weeks have been hell in terms of free time/energy.
Was taking a look at the ConcreteItem class and how I would go about implementing it, but find a spare 1-2 hrs of active thinking time just hasn't happened I'm afraid.

How've you got on with making the changes?
If you need any of my questionable logic explain, give me shout.
I'm really hoping to find some time this week to sit down and get back up to speed with this!

File: syncall/tw_caldav_utils.py (hunk 10 above)
Comment (Pete-Hamlin) [reply syncall/tw_caldav_utils.py:174]: I'm afraid this was just shamelessly copied from here: https://github.com/bergercookie/syncall/blob/master/syncall/tw_gcal_utils.py#L228

Not sure of the wider implications on a given sync, I guess it would create a new TW item if it can't find the associated UUID?

**Comment by bergercookie** [comment]
Hi @Pete-Hamlin ,

No prob :)

I think the PR is in a pretty good state, good job!
I did a few tests with a local [radicale](https://radicale.org/3.0.html) server of mine and it works nicely.

I want to test some edge cases and then it should be good to go! 🎉🎉🎉

File: taskwarrior_syncall/caldav_side.py [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+        self._calendar_name = calendar_name
+        self._calendar = None
+        self._items_cache: Dict[str, dict] = {}
+
+    def start(self):
+        logger.info(f"Initializing {self.fullname}...")
+        self._calendar = self._get_calendar()
+        logger.debug(f"Connected to calendar - {self._calendar.name}")
+
+    def _get_calendar(self) -> caldav.DAVClient.calendar:
```
[the remark this answers is not in the document]
Comment (bergercookie) [reply taskwarrior_syncall/caldav_side.py]: done

**Comment by bergercookie** [comment]
OK, one last commit from my side and I think it's ready to go!
Feel free to look at the commit message as I tried to be as descriptive as possible to shed some light into the inner workings of the implementation and why I made these changes.

if you do iterate further on the code, please go through this list, for the `tw_caldav_sync` executable at least:

* [ ] Verify you can add items from TW and they appear in caldav during the initial run with both sides being empty
* [ ] Verify you can add items from TW and they appear in caldav during the initial run with caldav being empty at first
* [ ] Verify you can add items from caldav and they appear in TW during the initial run with TW being empty at first
* [ ] Verify you can update the description of items from TW and the changes appear in caldav
* [ ] Verify you can update the description of items from caldav and the changes appear in TW
* [ ] Verify you can mark items as completed from TW and they get completed in caldav
* [ ] Verify you can mark items as completed from caldav and they get completed in TW
* [ ] Verify you can delete items from TW and they get deleted in caldav
* [ ] Verify you can delete items from caldav and they get deleted in TW
* [ ] Verify you can create *completed* items from TW and they appear completed in caldav
* [ ] Verify you can create *completed* items from caldav and they appear completed in TW
* [ ] Verify you can create *deleted* items from TW and they appear deleted in caldav
* [ ] Verify you can create *deleted* items from caldav and they appear deleted in TW
* [ ] Verify you can create in-progress items from TW and they appear as in-progress/started in caldav
* [ ] Verify you can create in-progress items from caldav and they appear as in-progress/started in TW

I understand it's rather repetitive but atm it's the only way to verify that this integration will work with most of the combinations users may throw at it. In the future it would be awesome to have some integration tests that verify the above automatically. Let me know if you don't plan to add more changes and we can skip the above.

Also, if you have some time feel free to add A GIF/video showcasing how the integration works - See for example the TW<>Notion integration. https://github.com/bergercookie/syncall/blob/master/misc/tw_notion_sync.gif. I recommend [gifski](https://gif.ski/) and [simplescreenrecorder](https://www.maartenbaert.be/simplescreenrecorder/) but not too fussed. If you don't have much time, happy to merge it as is.

**Comment by Pete-Hamlin** [comment]
Awesome, thank you for the updates! I'm 100% not going to add anything else at this point. If I do any future work/want to add anything, it'll probably be some time in the future when I have a little more time available.

All looks good to me! I'll try and do some confirmation testing from my side at lunch, and will update with a gif if possible :) 

Thanks once again for your input!

**Comment by Pete-Hamlin** [comment]
So I've tested it all at my end and all looks good.

I looked into adding a gif for it, but the only instance URL I have to test against is my personal one, which I'd prefer not to add to a public repo just for privacy reasons :/ Apologies, will have a think if I can setup something in a way that masks the URL.

**Comment by saccarosium** [comment]
Hi guys, 
I just wanna say that I've tested this PR on my Nextcloud and works good. @bergercookie if you need a GIF I can make one myself.

**Comment by bergercookie** [comment]
Hi, apologies, I was on holiday.


> I looked into adding a gif for it, but the only instance URL I have to test against is my personal one, which I'd prefer not to add to a public repo just for privacy reasons :/ Apologies, will have a think if I can setup something in a way that masks the URL.

Hmm, you can always take a snapshot of part of the screen - or specific windows like a terminal + a private tab. This way you don't have to screen capture your entire screen.

In any case, the GIF is not a dealbreaker. @saccarosium , @Pete-Hamlin if any of you could create one and send it over I could later upload it via a separate commit to the repo.

I'll try to test that these changes don't break the rest of the dependencies and hopefully I should be able to merge soon.

**Comment by bergercookie** [comment]
Merged it. Thanks @Pete-Hamlin  for the work 🎉

Do feel free to send me / create a new PR with the GIF.
